# sequence-design-guide
Protocol to redesign enzyme sequences with ProteinMPNN followed in Krasnow et. al. 2025

## Section I: generate residue constraints

### Distance analysis for active site constraints

This step will analyze the structure of your protein and measure the distance from each residue to the active site, generating list of constrained residues below a specified distance cutoff. Since it is likely that no structure of your protein has been solved in complex with substrate, the script is written to do the distance measurements on a homolog bound to substrate and map the constrained residues to your target protein. You can use a substrate-bound structure of the target protein if one is available or an AlphaFold3 model if you trust it.

Install required software: 
- distance_constraint_analysis.ipynb (Google Colab Notebook)

1.	Find a homolog of your target protein that is closely related in structure and has a structure solved with the substrate bound, or generate an AF3 structure. Download the pdb file of this substrate-bound structure.
2.	Create a folder in your Google Drive home directory titled ‘pdbs’. Upload the pdb file of the pdb of the homolog with substrate bound into this folder.
3.	Create a pairwise alignment between your protein sequence and the homolog protein sequence (this can be done in Geneious). The homolog should be the first sequence and the target protein should be the second. Export the alignment in .fasta format.
4.	Create a folder in your Google Drive home directory titled ‘fastas’. Upload the .fasta alignment from step 2.
5.	Upload the Colab notebook to your ‘Colab Notebooks’ folder to your Drive home directory.
6.	Run all the cells in the Colab Notebook one at a time, following instructions in the file as indicated in the text. Comments are provided to explain code and provide warnings for potential changes needed for different situations that weren't encounted in this paper. 

### Conservation anaylsis for evolutionary constraints

This step will conduct a sequence homology search to analyze the conservation of each residue. This information will be used after to generate lists of residues to constrain from design.

Required software: 

- BLAST+ (installation available at https://ftp.ncbi.nlm.nih.gov/blast/executables/LATEST/ and documentation available at https://www.ncbi.nlm.nih.gov/books/NBK279690/). 
- Python3 https://www.python.org/downloads/. Python pachages: biopython, pandas (install with pip with the terminal command e.g., pip install biopython)
- Jupyter notebook (https://jupyter.org/install)
- Conservation_and_constraints.ipynb 

BLAST+ Installation instructions provided below. For Python and Jupyter, follow instructions on the links provided.

1.	Download the executable file from the above link for your operating system (e.g., ncbi-blast-2.15.0+-x64-macosx.tar.gz. Double click to unzip and move the unzipped file to your documents   
2.	Add the program files to the Path so that your computer knows where to find them. Run the commands in the terminal:

`Echo PATH=$PATH:~/Documents/ncbi-blast-2.15.0+/bin>>.bashrc`
`Echo export PATH>>.bashrc`
`Source .bashrc`

Test the installation by entering 
`Blastp`
If this returns a blast-related error other than “command not found” then the installation was successful.

3.	Make a folder in your Documents where you’ll run blast analysis, name this folder blast_analysis. Make another folder in your Documents where you’ll store the database you’re searching, name this folder “dbs” (this folder should not be nested within blast_analysis)
4.	Put a .fasta file of your target protein sequence in the blast_analysis folder. The fasta file can be generated by exporting from Geneious, downloading from unprot, etc., and it will serve as the query for the homology search.
5.	Download the database in which you want to conduct the homology search. I used uniref50: https://www.uniprot.org/help/downloads. Unzip and move the downloaded fasta file to the dbs folder
6.	Convert the fasta file into a BLAST-searchable database with the terminal commands:
`cd ~/Documents/dbs`

`Makeblastdb -in uniref50.fasta -title uniref50 -dbtype prot -out uniref50 -parse_seqids`

8.	Run the blast search for your target protein against your chosen database with the terminal commands, replacing *.fasta with the name of your query sequence fasta file:

`cd ~/Documents/blast_analysis`
`blastp -query *.fasta -db ../dbs/uniref50 -out blast_out_table.tsv -evalue 1e-4 -outfmt 6`
A file named “blast_out_table.csv” will be generated which contains a list of homologs found. The -evalue flag value can be tuned to modulate the stringency of the search (smaller evalue = more stringent).
8.	Extract the sequences of the search hits: open the blast_out_table.csv file and copy the “ids” column values into a text file called hits.txt in the same folder (don’t include the column header). Run the terminal command:
`Blastdbcmd -db ../dbs/uniref50 -dbtype prot -entry_batch hits.txt -outfmt %f -out hits.fasta`
The file hits.fasta containing the sequence of each hit from the search should have been created. These search results will be used in the next step to quantify sequence conservation and generate the list of residues to constrain by conservation. 

9.	Copy the conservation_and_constraints.ipynb file to the directory that has the Blast anaylsis outputs
10.	Open Jupyter notebook with the terminal command: jupyter notebook. When jupyter opens, navigate to the folder with the .ipynb file and open it
11.	Run the cells one at a time, pasting in the names your target protein fasta and Blast output files where needed. Paste in the distance-constrained residue list when prompted. 
12.	Check the ‘_constraints.csv’ output generated. Confirm that the correct residues were selected using the pymol commands. The ‘MPNN fix string’ column will be used in the next step, which is to predict protein sequences using ProteinMPNN.

